### PR TITLE
HESA NUTS3 and LEP

### LINKED ISSUE #254: HESA data to NUTS3 and LEPs

### PR DESCRIPTION
**Pull request by georgerichardson**
- Add NUTS3 and LEP indicators for HESA
- Add NUTS3 and LEP schemas for HESA
- Replace previous reverse geocoding and indicator creation utils with new functions that can handle overlapping regions (for LEP compatibility)

closes #254

### CODE REVIEW

File: ds/data/processed/hesa/area_university_site.nuts3.csv [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+2015,UKC14,2013,200.0
+2016,UKC14,2013,200.0
+2017,UKC14,2013,258.0
+2018,UKC14,2016,258.0
+2015,UKC22,2013,680.854
+2016,UKC22,2013,677.064
+2017,UKC22,2013,678.223
+2018,UKC22,2016,629.923
+2015,UKC23,2013,29.72
+2016,UKC23,2013,28.01
```
Comment (mindrones) [inline ds/data/processed/hesa/area_university_site.nuts3.csv:15]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/area_university_site.nuts2.csv [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+2015,UKJ3,2013,152.62199999999999
+2016,UKJ3,2013,144.002
+2017,UKJ3,2013,143.03199999999998
+2018,UKJ3,2016,214.00799999999998
+2015,UKJ4,2013,147.95
+2016,UKJ4,2013,152.70000000000002
+2017,UKJ4,2013,152.002
+2018,UKJ4,2016,151.699
+2015,UKK1,2013,558.5129999999999
+2016,UKK1,2013,523.566
```
Comment (mindrones) [inline ds/data/processed/hesa/area_university_site.nuts2.csv:119]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/fte_research_students.lep.csv [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+2015,E37000005,2014,20.0
+2016,E37000005,2014,15.0
+2017,E37000005,2017,20.0
+2018,E37000005,2017,25.0
+2015,E37000006,2014,1825.0
+2016,E37000006,2014,1840.0
+2017,E37000006,2017,1925.0
+2018,E37000006,2017,1935.0
+2015,E37000007,2014,50.0
+2016,E37000007,2014,45.0
```
Comment (mindrones) [inline ds/data/processed/hesa/fte_research_students.lep.csv:27]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/fte_research_students.nuts2.csv [hunk 4]
```diff
@@ -0,0 +1,161 @@
+year,nuts_id,nuts_year_spec,fte_research_students
+2015,UKC1,2013,1355.0
+2016,UKC1,2013,1335.0
+2017,UKC1,2013,1360.0
+2018,UKC1,2016,1365.0
+2015,UKC2,2013,2335.0
+2016,UKC2,2013,2390.0
```
Comment (mindrones) [inline ds/data/processed/hesa/fte_research_students.nuts2.csv:7]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/fte_research_students.nuts3.csv [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+2015,UKC22,2013,2200.0
+2016,UKC22,2013,2270.0
+2017,UKC22,2013,2245.0
+2018,UKC22,2016,2230.0
+2015,UKC23,2013,135.0
+2016,UKC23,2013,120.0
+2017,UKC23,2013,145.0
+2018,UKC23,2016,140.0
+2015,UKD12,2013,50.0
+2016,UKD12,2013,45.0
```
Comment (mindrones) [inline ds/data/processed/hesa/fte_research_students.nuts3.csv:19]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/gbp_research_income.nuts2.csv [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+2015,UKC2,2013,159301000.0
+2016,UKC2,2013,156938000.0
+2017,UKC2,2013,159548000.0
+2018,UKC2,2016,162619000.0
+2015,UKD1,2013,531000.0
+2016,UKD1,2013,610000.0
+2017,UKD1,2013,806000.0
+2018,UKD1,2016,655000.0
+2015,UKD3,2013,366059000.0
+2016,UKD3,2013,356437000.0
```
Comment (mindrones) [inline ds/data/processed/hesa/gbp_research_income.nuts2.csv:15]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/gbp_research_income.nuts3.csv [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+2015,UKC14,2013,76311000.0
+2016,UKC14,2013,78575000.0
+2017,UKC14,2013,77807000.0
+2018,UKC14,2016,78905000.0
+2015,UKC22,2013,156273000.0
+2016,UKC22,2013,154157000.0
+2017,UKC22,2013,156678000.0
+2018,UKC22,2016,159426000.0
+2015,UKC23,2013,3028000.0
+2016,UKC23,2013,2781000.0
```
Comment (mindrones) [inline ds/data/processed/hesa/gbp_research_income.nuts3.csv:15]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_postgraduates.nuts2.csv [hunk 8]
```diff
@@ -0,0 +1,198 @@
+year,nuts_id,nuts_year_spec,total_postgraduates
+2014,UKC1,2010,1260.0
+2015,UKC1,2013,1255.0
+2016,UKC1,2013,1275.0
+2017,UKC1,2013,1265.0
+2018,UKC1,2016,1285.0
+2014,UKC2,2010,2130.0
+2015,UKC2,2013,2140.0
+2016,UKC2,2013,2255.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_postgraduates.nuts2.csv:9]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_postgraduates.nuts3.csv [hunk 9]
```diff
@@ -0,0 +1,432 @@
+year,nuts_id,nuts_year_spec,total_postgraduates
+2014,UKC12,2010,75.0
+2015,UKC12,2013,85.0
+2016,UKC12,2013,135.0
+2017,UKC12,2013,145.0
+2018,UKC12,2016,150.0
+2014,UKC14,2010,1185.0
+2015,UKC14,2013,1170.0
+2016,UKC14,2013,1140.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_postgraduates.nuts3.csv:9]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_stem_postgraduates.nuts2.csv [hunk 10]
```diff
@@ -0,0 +1,198 @@
+year,nuts_id,nuts_year_spec,total_stem_postgraduates
+2014,UKC1,2010,575.0
+2015,UKC1,2013,570.0
+2016,UKC1,2013,595.0
+2017,UKC1,2013,640.0
+2018,UKC1,2016,650.0
+2014,UKC2,2010,1010.0
+2015,UKC2,2013,1045.0
+2016,UKC2,2013,1040.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_stem_postgraduates.nuts2.csv:9]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_stem_postgraduates.nuts3.csv [hunk 11]
```diff
@@ -0,0 +1,432 @@
+year,nuts_id,nuts_year_spec,total_stem_postgraduates
+2014,UKC12,2010,45.0
+2015,UKC12,2013,45.0
+2016,UKC12,2013,55.0
+2017,UKC12,2013,70.0
+2018,UKC12,2016,65.0
+2014,UKC14,2010,530.0
+2015,UKC14,2013,525.0
+2016,UKC14,2013,540.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_stem_postgraduates.nuts3.csv:9]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_stem_students.nuts2.csv [hunk 12]
```diff
@@ -0,0 +1,198 @@
+year,nuts_id,nuts_year_spec,total_stem_students
+2014,UKC1,2010,9610.0
+2015,UKC1,2013,9845.0
+2016,UKC1,2013,10335.0
+2017,UKC1,2013,10570.0
+2018,UKC1,2016,11080.0
+2014,UKC2,2010,15730.0
+2015,UKC2,2013,15990.0
+2016,UKC2,2013,16215.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_stem_students.nuts2.csv:9]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_stem_students.nuts3.csv [hunk 13]
```diff
@@ -0,0 +1,432 @@
+year,nuts_id,nuts_year_spec,total_stem_students
+2014,UKC12,2010,4380.0
+2015,UKC12,2013,4520.0
+2016,UKC12,2013,4915.0
+2017,UKC12,2013,4945.0
+2018,UKC12,2016,5140.0
+2014,UKC14,2010,5230.0
+2015,UKC14,2013,5325.0
+2016,UKC14,2013,5420.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_stem_students.nuts3.csv:9]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_university_buildings.lep.csv [hunk 14]
```diff
@@ -0,0 +1,151 @@
+year,lep_id,lep_year_spec,total_university_buildings
+2015,E37000001,2014,84.0
+2016,E37000001,2014,84.0
+2017,E37000001,2017,86.0
+2018,E37000001,2017,88.0
+2015,E37000002,2014,27.0
+2016,E37000002,2014,27.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_university_buildings.lep.csv:7]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_university_buildings.nuts2.csv [hunk 15]
```diff
@@ -0,0 +1,161 @@
+year,nuts_id,nuts_year_spec,total_university_buildings
+2015,UKC1,2013,407.0
+2016,UKC1,2013,406.0
+2017,UKC1,2013,408.0
+2018,UKC1,2016,472.0
+2015,UKC2,2013,361.0
+2016,UKC2,2013,348.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_university_buildings.nuts2.csv:7]: isn't `nuts_year_spec` here supposed to be 2016?

File: ds/data/processed/hesa/total_university_buildings.nuts3.csv [hunk 16]
```diff
@@ -0,0 +1,357 @@
+year,nuts_id,nuts_year_spec,total_university_buildings
+2015,UKC12,2013,94.0
+2016,UKC12,2013,94.0
```
Comment (mindrones) [inline ds/data/processed/hesa/total_university_buildings.nuts3.csv:3]: isn't `nuts_year_spec` here supposed to be 2016?

**Review by mindrones — COMMENTED** [review]
It would seem good apart from NUTS `year`/`nuts_year_spec` discrepancies (I don't know LEP's spec years, so I haven't commented on `.lep.csv` files)

**Review by mindrones — DISMISSED** [review]
I forgot to mention that for indicators that will be visualised in the tool, we need to add a log in `CHANGELOG.md` before merging (this file is linked in the tool so the language should not be too technical)

**Comment by georgerichardson** [comment]
For the `nuts_year_spec`: NUTS 2016 is enforced from [2018 onwards](https://ec.europa.eu/eurostat/web/nuts/history)

Thanks for the heads up about the changelog. In this case no new NUTS2 data has been added (I've just changed the file names from `.csv` to `.nuts2.csv`. Should I still include an update?

**Review by mindrones — APPROVED** [review]
> For the nuts_year_spec: NUTS 2016 is enforced from 2018 onwards

OK we discussed about this in Slack and decided to change it in one go (#301)

> no new NUTS2 data has been added (I've just changed the file names from .csv to .nuts2.csv. Should I still include an update?

Ah then no, approving for my part 👍

**Pull request merged**